### ISSUE TITLE
`$this` incorrectly detected inside object creation expression

### ISSUE THREAD

**Issue by przepompownia** [phpactor/phpactor#2618]
![image](https://github.com/phpactor/phpactor/assets/11404453/20383f72-fc84-404a-93d9-399a0d1cca01)

The case
```php
        return new Success((function () {
            $classes = $this;
        })());
```
exists for example in `lib/CodeTransform/Adapter/WorseReflection/Transformer/ImplementContracts.php`. 

It's a regression after #2469 @mamazu 

Perhaps similar caution will be needed in #2613

**Comment by mamazu** [comment]
Could you maybe come up with a test case for that and I'll have a look at it.

**Comment by przepompownia** [comment]
https://github.com/phpactor/phpactor/pull/2619

**Comment by przepompownia** [comment]
I also looked into `FunctionLikeWalker` but don't have idea how to distinguish whether created object belongs to named or anonymous class.

**Comment by mamazu** [comment]
The distinguishing feature is that the `typeDescrimitator` of the `ObjectCreationExpression` is a Token and not a `Node`. That's what we are using in the other PR.

**Comment by mamazu** [comment]
I've created a fix pr: @2620

I still hold the opinion that the Parser should return a more coherent structure but oh well.

**Comment by przepompownia** [comment]
Fixed by #2620
